README.md
# users-api

A small example homework application

## Usage

### CLI Example

    lein run -m users-api.cli --pipes-file a.psv \
      --commas-file a.csv \
      --pipes-file b.psv \
      --spaces-file a.ssv \
      --sort birthdate

This will read in the given files, parse them appropriately, and print the
output sorted by date of birth

### Web API example

To start the API, simply do

    lein ring server

The location of a seed data file can be configured in `resources/config.edn`

To use the API, you can do

    curl -H 'content-type: text/pipe-separated' -X POST  http://localhost:3000/records -d 'Brown|Jonathan |M| Green | 1988-01-18'

to insert a pipe-separated record. Other supported content types are `text/csv`
and `text/space-separated`.

To query records, the same API supports `/records/name`, `/records/birthdate`,
and `/records/gender`. These endpoints will return JSON.

## License

Copyright © 2019 Jonathan Brown

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-04-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-04-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/users-api/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/users-api/compare/0.1.0...0.1.1


test/users_api/cli_test.clj
(ns users-api.cli-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :as io]
            [users-api.cli :refer :all]
            [clojure.tools.cli :refer [parse-opts]])
  (:import [java.time LocalDate]))

(deftest options-test
  (testing "options are parsed correctly"
    (let [args ["--pipes-file" "foo.txt"
                "--pipes-file" "bar.txt"
                "--spaces-file" "baz.txt"
                "--sort" "gender"]]
      (is (= (:options (parse-opts args cli-options))
             {:pipes-file ["foo.txt" "bar.txt"]
              :spaces-file ["baz.txt"]
              :commas-file []
              :sort-order :gender}))))

  (testing "Error is present for invalid options"
    (let [args ["--pipes-file" "foo.txt"
                "--sort" "height"]]
      (is (= (:errors (parse-opts args cli-options))
             ["Failed to validate \"--sort height\": Please choose one of \"dob\", \"gender\" \"last-name\" "])))))

(deftest str-test
  (testing "Record is serialized correctly"
    (let [record {:last-name "BBB", :first-name "FirstName",
                  :gender :female, :favorite-color "Orange",
                  :dob (LocalDate/parse"1985-11-20")}]
      (is (= (user-str record)
             "BBB, FirstName, female, Orange, 11/20/1985")))))


test/users_api/web_test.clj
(ns users-api.web-test
  (:require [clojure.test :refer :all]
            [clojure.java.io :as io]
            [users-api.web :refer :all]
            [cheshire.core :refer [parse-string]])
  (:import [java.time LocalDate]))

(def test-rows
  [{:last-name "aaa", :first-name "FirstName", :gender :female, :favorite-color "Red",
    :dob (LocalDate/parse "1992-02-02")}
   {:last-name "Brown", :first-name "Jonathan", :gender :male, :favorite-color "Green",
    :dob (LocalDate/parse "1988-01-18")}
   {:last-name "zAz", :first-name "FirstName", :gender :female, :favorite-color "Orange",
    :dob (LocalDate/parse"1985-11-10")}
   {:last-name "BBB", :first-name "FirstName", :gender :female, :favorite-color "Orange",
    :dob (LocalDate/parse"1985-11-11")}])

(deftest init-data-test
  (testing "data is loaded"
    (do
      (reset! records [])
      (init-data)
      (is (= (map :favorite-color @records)
             ["Red" "Green" "Orange" "Orange"]))
      (reset! records []))))

;; Should really use ring-mock for these, and possibly clojure.test/use-fixtures
(deftest routes-test
  (testing "ping"
    (is (= (:body (app {:uri "/ping"
                        :request-method :get}))
           "pong")))

  (testing "get records"
    (do
      (reset! records test-rows)
      (is (= (map :last-name
              (-> (app {:uri "/records/gender" :request-method :get})
                  :body slurp (parse-string true)))
             ["aaa" "BBB" "zAz" "Brown"]))

      (is (= (map :dob
              (-> (app {:uri "/records/birthdate" :request-method :get})
                  :body slurp (parse-string true)))
             ["1985-11-10" "1985-11-11" "1988-01-18" "1992-02-02"]))

      (is (= (map :last-name
              (-> (app {:uri "/records/name" :request-method :get})
                  :body slurp (parse-string true)))
             ["zAz" "Brown" "BBB" "aaa"]))))

  (testing "add a record"
    (do
      (reset! records [])
      (is (= (:status (app {:uri "/records" :request-method :post
                            :content-type "text/space-separated"
                            :headers {"content-type" "text/space-separated"}
                            :body (-> "Hawk Tony M Red 1968-05-12" (.getBytes) io/input-stream)}))
             201))
      (is (= (-> @records first :dob)
             (LocalDate/parse "1968-05-12"))))))


test/users_api/core_test.clj
(ns users-api.core-test
  (:require [clojure.test :refer :all]
            [users-api.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/users_api/cli.clj
(ns users-api.cli
  (:require [clojure.tools.cli :refer [parse-opts]]
            [clojure.string :as str]
            [users-api.data :as data])
  (:import [java.time.format DateTimeFormatter]))

(def cli-options
  [["-c" "--commas-file PATH" "Path to a CSV data file"
    :default []
    :assoc-fn (fn [m k v]
                (update-in m [k] conj v))
    :id :commas-file]

   ["-s" "--spaces-file PATH" "Path to a space-separated data file"
    :default []
    :assoc-fn (fn [m k v]
                (update-in m [k] conj v))
    :id :spaces-file]

   ["-p" "--pipes-file PATH" "Path to a pipe-separated data file"
    :default []
    :assoc-fn (fn [m k v]
                (update-in m [k] conj v))
    :id :pipes-file]

   ["-z" "--sort ORDER" "Key by which to sort the results"
    :parse-fn keyword
    :default :name
    :validate [#{:birthdate :gender :name}
               "Please choose one of \"dob\", \"gender\" \"last-name\" "]
    :id :sort-order]])

(defn user-str [user]
  (let [u (-> user
              (update :dob #(.format % (DateTimeFormatter/ofPattern "M/d/uuuu")))
              (update :gender name))]
   (str/join ", " (for [k data/user-keys] (get u k)))))

(defn -main [& args]
  (let [opts (parse-opts args cli-options)]
    (if-let [e (:errors opts)]
      (prn e)
      (let [rows (data/data-seq (concat
                                 (map #(vector data/pipe-sep %)
                                      (get-in opts [:options :pipes-file]))
                                 (map #(vector data/space-sep %)
                                      (get-in opts [:options :spaces-file]))
                                 (map #(vector data/comma-sep %)
                                      (get-in opts [:options :commas-file]))))

            sort-fn (get data/sort-orders (get-in opts [:options :sort-order]))]
        (doseq [r (sort-fn rows)]
          (prn (user-str r)))))))


src/users_api/data.clj
(ns users-api.data
  (:require [clojure.string :as str]
            [clojure.java.io :as io])
  (:import [java.time LocalDate]))


(defn lazy-line-seq
  "Returns a lazy seq of lines from the given reader. Closes the reader
  once it has been consumed"
  [rdr]
  (lazy-seq (if-let [next (.readLine rdr)]
              (cons next (lazy-line-seq rdr))
              (.close rdr))))

(def user-keys [:last-name :first-name :gender :favorite-color :dob])

(defn row->rowmap
  "converts a string into a map of keywords and parsed values"
  [separator row]
  (let [keys user-keys
        vals (str/split row separator)]
    (-> (zipmap keys vals)
        (update :gender #(condp contains? (str/lower-case %)
                           #{"f" "female" "woman"} :female
                           #{"m" "male" "man"}     :male
                           :other))
        (update :dob #(LocalDate/parse (str/trim %))))))

(defn path->rows [separator filename]
  (->> filename io/reader lazy-line-seq
       (map (partial row->rowmap separator))))

(defn data-seq
  "`files` should be a seq of vectors that look like [separator, path]"
  [files]
  (mapcat #(apply path->rows %) files))

(def sort-orders {:gender (partial sort-by (juxt :gender (comp str/lower-case :last-name)))
                  :name   #(reverse (sort-by (comp str/lower-case str/trim :last-name) %))
                  :birthdate    (partial sort-by :dob)})

;; regexes for splitting strings by the appropriate separator
(def pipe-sep #" *\| *")
(def space-sep #" +")
(def comma-sep #" *, *")


src/users_api/web.clj
(ns users-api.web
  (:require [compojure.api.sweet :refer [defapi context GET POST undocumented]]
            [ring.util.http-response :refer [ok created bad-request! not-found]]
            [ring.middleware.defaults :refer [wrap-defaults api-defaults]]
            [compojure.route]
            [schema.core :as s]
            [users-api.data :as data]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [clojure.edn :as edn]))

;; should probably use something like aero for this
(def config (edn/read-string (slurp (io/resource "config.edn"))))

(def records (atom []))

;; TODO validate input with schema/spec
(defn insert! [separator row]
  (swap! records conj (data/row->rowmap separator row)))

(def separators
  {"text/csv" data/comma-sep
   "text/pipe-separated" data/pipe-sep
   "text/space-separated" data/space-sep})

(defapi app
  (context "/" []
           (GET "/ping" []
                (ok "pong"))

           (GET "/records/:sort-order" [sort-order]
                (let [sort-fn (get data/sort-orders (keyword sort-order))]
                  (ok (sort-fn @records))))

           (POST "/records" req
                 (let [created-resp {:status 201 :headers {} :body "created"}
                       body (-> req :body slurp)]
                   (if-let [separator (get separators (:content-type req))]
                     (do body
                       (insert! separator body)
                       (created))
                     (bad-request! "Invalid content-type"))))
           (undocumented
            (compojure.route/not-found (ok {:not "found"})))))

(defn init-data []
  (let [{:keys [separator seed-file]} (:data config)
        initial-data (->> seed-file io/resource slurp str/split-lines
                          (map (partial data/row->rowmap (re-pattern separator))))]
    (reset! records (vec initial-data))))

;; launch a dev server in a repl
(comment
  (require '[ring.adapter.jetty :refer [run-jetty]])
  (require '[ring.middleware.reload :refer [wrap-reload]])
  (run-jetty (wrap-reload #'app) {:port 3000 :join? false})
  @records
  )
